### DISCUSSION TITLE
About audio output（StelAudioMgr）

### DISCUSSION THREAD

**Discussion by Fuyuko-SFA** [Stellarium/stellarium#3360] [Q&A] [answered]
Hi Everybody,

Is ”StelAudioMgr" working with Stellarium 23.1 ?

The script AudioTest.ssc is written as under:

> StelAudioMgr.loadSound("../work/Neighing_horse.mp3","Horse");
> StelAudioMgr.playSound("Horse");

But this cannot play sound.

The folder structure is below. 

- stellarium.exe = C:\Program Files (x86)\Stellarium\stellarium.exe
- Script (AudioTest.ssc) = C:\Program Files (x86)\Stellarium\scripts\AudioTest.ssc
- Audio File ( Neighing_horse.mp3 ) = C:\Program Files (x86)\Stellarium\work\Neighing_horse.mp3

Any help or advice would be much appreciated.
Thanks in advance!

Fuyuko

**Comment by gzotti** [comment]
I just ran the audiotest script in 23.2. The OGG is missing, WAV and MP3 play well. In master, OGG plays but WAV sounds bad.
 
Two suggestions, not sure if any applies here:
- you should NEVER place your own data into the Stellarium program folder. Use C:\Users\YOU\AppData\Roaming\stellarium as described in the User Guide and everywhere else. 
- Make sure you have audio codecs installed. Does your file play with other programs?

Looking again, it seems the path is likely a problem. the Audio files must reside in the script directory or a subdirectory of that.

**Comment by Fuyuko-SFA** [comment]
Hi,
Thank you @gzotti

I moved mp3 from "C:\Program Files (x86)\Stellarium\work\" to  "C:\Users\Fuyuko\AppData\Roaming\Stellarium\work"
Then I modified the AudioTest.ssc as under:

> StelAudioMgr.loadSound("C:\Users\Fuyuko\AppData\Roaming\Stellarium\work\Neighing_horse.mp3", "Horse");
> StelAudioMgr.playSound("Horse");

But this cannot play sound.
Also I moved mp3 to the script directory (C:\Program Files (x86)\Stellarium\scripts\Neighing_horse.mp3)
And I modified the AudioTest.ssc as under:

> StelAudioMgr.loadSound("Neighing_horse.mp3", "Horse");
> StelAudioMgr.playSound("Horse");

But this cannot play sound either.

Neighing_horse.mp3 can play with Windows Media Player.

Do I need any plugins to use "StelAudioMgr"?
Is the script I wrote correct?

Any other ideas for playing audio?

Many thanks
Fuyuko

**Comment by alex-w** [comment]
Please use relative paths for sounds

**Comment by gzotti** [comment, marked as answer]
> Hi, Thank you @gzotti
> 
> I moved mp3 from "C:\Program Files (x86)\Stellarium\work" to "C:\Users\Fuyuko\AppData\Roaming\Stellarium\work" Then I modified the AudioTest.ssc as under:
> 
> > StelAudioMgr.loadSound("C:\Users\Fuyuko\AppData\Roaming\Stellarium\work\Neighing_horse.mp3", "Horse");

Absolute path names are forbidden for reasons of security.
Place the MP3 into C:\Users\Fuyuko\AppData\Roaming\Stellarium\scripts\work, then try
StelAudioMgr.loadSound("work/Neighing_horse.mp3", "Horse");

If that does not work, try our officially supported call by the core object which we use in our audiotest.ssc:

core.loadSound("work/Neighing_horse.mp3", "Horse");


> > StelAudioMgr.playSound("Horse");
> 
> But this cannot play sound. Also I moved mp3 to the script directory (C:\Program Files (x86)\Stellarium\scripts\Neighing_horse.mp3) 

You did it again. Don't put your stuff into the program directory. 

> And I modified the AudioTest.ssc as under:
> 
> > StelAudioMgr.loadSound("Neighing_horse.mp3", "Horse");
> > StelAudioMgr.playSound("Horse");
> 
> But this cannot play sound either.
> 
> Neighing_horse.mp3 can play with Windows Media Player.
> 
> Do I need any plugins to use "StelAudioMgr"?

No.

> Any other ideas for playing audio?

Check the logfile for any error messages.

**Comment by Fuyuko-SFA** [comment]
Thank you @gzotti

I have tried replacing the Neighing_horse.mp3 with the below file
＃ https://github.com/Stellarium/stellarium/tree/master/scripts/tests/audiotest.mp3

I also modified the script to use "core.loadSound".

Then audiotest.mp3 could play sound!
Maybe there's something wrong with theNeighing_horse.mp3.
I check Neighing_horse.mp3.

Thank you so much for the answers! 
Fuyuko

**Comment by ImmersiveTheatres** [comment]
Please may I ask the same question as @Fuyuko-SFA to @gzotti and / or @alex-w : but for Stellarium 23.3 running on macOS 14.1.2 

**Is StelAudioMgr functional in Stellarium 23.3?**

I have placed my audio file in: 

/Users/Me/Library/Application Support/Stellarium/scripts/audio/Polorum.mp3

ALSO in:

/Users/Me/Library/Application Support/Stellarium/scripts/Polorum.mp3

ALSO in:

/Applications/Stellarium.app/Contents/Resources/scripts/audio/Polorum.mp3


I have tried to load this audio file - both via the Script Console and a script file - using each of the following commands:

StelAudioMgr.loadSound("Me/Library/Application Support/Stellarium/scripts/audio/Polorum.mp3","Chant");
StelAudioMgr.loadSound("Applications/Stellarium.app/Contents/Resources/scripts/audio/Polorum.mp3","Chant");
StelAudioMgr.loadSound("audio/Polorum.mp3","Chant");
StelAudioMgr.loadSound("Polorum.mp3","Chant");

coreloadSound("Me/Library/Application Support/Stellarium/scripts/audio/Polorum.mp3","Chant");
coreloadSound("Applications/Stellarium.app/Contents/Resources/scripts/audio/Polorum.mp3","Chant");
coreloadSound("audio/Polorum.mp3","Chant");
coreloadSound("Polorum.mp3","Chant");

Yet none of them work. The same error is always returned:

script error: 'Reference'  @ line 1: ReferenceError: StelAudioMgr is not defined
OR
script error: 'Reference'  @ line 1: ReferenceError: coreloadSound is not defined


Can someone please help / advise?

Thanks very much

Mario

**Reply by alex-w** [reply]
Please use relative paths

P.S. Where is your script located?

**Reply by gzotti** [reply]
The error message is "StelAudioMgr is not defined". Is the Mac version without media support?

**Reply by ImmersiveTheatres** [reply]
I think I have above? I've tried all different options?

**Reply by ImmersiveTheatres** [reply]
What is media support, sorry I'm not familiar with that?

**Reply by ImmersiveTheatres** [reply]
The scripts are in: Applications/Stellarium.app/Contents/Resources/scripts/

**Reply by gzotti** [reply]
On second view: coreloadSound() must actually be core.loadSound()

**Reply by gzotti** [reply]
And yet again, you should not edit inside the app directory with elevated privileges (root) but in your user area.

**Reply by ImmersiveTheatres** [reply]
I see above you suggest:

"Make sure you have audio codecs installed. Does your file play with other programs?"

I have the latest MacBook Pro (purchased last week). Does it require additional media resources?

**Reply by alex-w** [reply]
> Is the Mac version without media support?

I built it with media support.

**Reply by gzotti** [reply]
I don't know. Again,  _Does your file play with other programs on this system?_

**Reply by alex-w** [reply]
I can test a simple media script on my Mac tomorrow (here is night right now and I prepare to sleep)

**Comment by ImmersiveTheatres** [comment]
Blimey, not sure why I used core.loadSound() incorrectly (!!)

Now it loads with core.loadSound

I assume I will now run:  core.playSound("Chant")

**Comment by ImmersiveTheatres** [comment]
Cool, it accepts:  core.playSound("Chant")

Yet I can't hear anything.

Is there a volume command I need maybe?

**Reply by alex-w** [reply]
Please try put a script and media file in same directory and use relative path to media file within script

**Comment by ImmersiveTheatres** [comment]
Yes the file plays with other players. It's OK, I will try different audio files, and will let you know. At least now the load command is working! THANK YOU guys! 😃

**Comment by ImmersiveTheatres** [comment]
IT WORKS!!!!

I have literally been waiting for YEARS for this functionality!! 😁 🤩 🥳

Thank you VERY MUCH @gzotti and @alex-w !! 🙏🏻 🙏🏻 🙏🏻🙏🏻🙏🏻

**Comment by gzotti** [comment]
I have (re)-activated this in 2016 for an exhibition project... Enjoy!
What was the final problem you did not hear it before?

**Reply by ImmersiveTheatres** [reply]
10 years ago I left the fixed planetarium industry and started working for myself with inflatable planetariums. I have always used Stellarium as the astronomy application for my presenter-led shows - yet in 2013-2015 there was no audio loading functionality, so I gave up on it, and have been playing the music manually for 10 years! 😂

It's great that I no longer have to play the music manually 😄

**Reply by ImmersiveTheatres** [reply]
Of course, the other challenge is that fixed planetarium operating systems (eg. Digistar, SkySkan, Zeiss, etc) are very easy to use / programme. I am not a coder / computer guy (most planetarium managers are science communicators / teachers). So it has been VERY difficult for me to understand the Stellarium coding.

Yet I have learned a lot over the past 6 months - thanks to @gzotti and @alex-w  😊

**Reply by gzotti** [reply]
I have first used Stellarium in a small (inflatable) dome only earlier this year and only recently had a first rather experimental show in a big dome. I had hoped to just use the Spout screen streaming, but this would require a large primary screen of the same resolution as the projection system. I therefore operated Stellarium with the RemoteControl web interface and the ancient Text User Interface (which could receive again some updates). Yes, some aspects or functions are not optimized for dome use, just because no developer works routinely in a dome. The scripting interface is commonly used to run complete scripts, not to enter interactive commands in a shell-like environment. There are also no interactive media controls, like video player position or volume control. This also explains why Stellarium's interface certainly does not look like the interfaces of the classical manufacturers - we just don't know more details than I have superficially seen the controls since the 1980s: many buttons and analog dials. In the digital era it's less dials, but a few screens and some proprietary command language, and occasional "in-house solutions" mixing commercial programs.

**Reply by ImmersiveTheatres** [reply]
It's OK though. Stellarium is AWESOME for inflatable planetariums. Most of my customers are primary schools, so I simply need basic functionality to impress my audiences.

I appreciate the fact you guys have included the following optional requests in the past, which really help to deliver a great inflatable dome experience:

[astro]
flag_forced_meteor_activity = true  [for planetariums to enable a show of sporadic meteors without atmosphere]

[stars]
flag_forced_twinkle  =  true    [for planetariums to enable twinkling of stars without atmosphere]

[gui]
flag_show_gui  =  false    [to completely hide the vertical & horizontal toolbars]

**Reply by gzotti** [reply]
Ah, that was you. The atmosphere is IMO important. You can tweak extinction and light pollution to show more stars (if that was the original problem), but twilight colors (esp. with the new model) are the best you can get.

**Comment by ImmersiveTheatres** [comment]
The biggest problem I am facing now (with v23.3 and a brand new MacBook Pro M3), is the sky motion is not smooth, but jerky.

I am currently researching this, and will be posting to this thread later: https://github.com/Stellarium/stellarium/issues/2947 (in case someone can give me advice on how to improve the performance)
